jujugui/static/gui/src/app/app.js

```diff
@@ -1011,11 +1011,82 @@ YUI.add('juju-gui', function(Y) {
     },
 
     /**
+      Get user name from cookie.
+
+      @method _getUsernameFromCookie
+    */
+    _getUsernameFromCookie: function() {
+      // XXX jcsackett 2016-01-08: We should be using the whoami endpoint on
+      // charmstore to get the username of a logged in jem user rather than
+      // parsing the cookie. Consider this all a bad hack for demo and we'll do
+      // something better in the very near future.
+
+      // Don't look for cookies if in sandbox.
+      if (this.get('sandbox')) {
+        return null;
+      }
+      var cookie = document.cookie;
+      var data = cookie.split(';');
+      var macaroon = null;
+      var prefix = 'Macaroons-jem=';
+      var found = data.some(function(item) {
+        macaroon = item;
+        return (item.indexOf(prefix) !== -1);
+      });
+      if (!found) {
+        console.error('No macaroon found to get username from.');
+        return null;
+      }
+      var index = macaroon.indexOf(prefix) + prefix.length;
+      macaroon.slice(index);
+      macaroon = macaroon.replace('Macaroons-jem=', '');
+      macaroon = JSON.parse(atob(macaroon));
+      var username = null;
+      // We check all the caveats of each piece of the macaroon for the username
+      // data. If we find more than one username we abandon the macaroon as
+      // invalid.
+      macaroon.forEach(function(piece) {
+        var caveats = piece.caveats;
+        caveats.forEach(function(caveat) {
+          var declaration = 'declared username ';
+          if (caveat.cid.indexOf(declaration) !== -1) {
+            if (username === null) {
+              username = caveat.cid.replace(declaration, '');
```

Comment:
Yeah, that could be better if we run into the same arbitrary whitespace issue as in the initial cookie parsing.